From: Stanislav Malyshev <stas@php.net>
Date: Mon, 27 May 2019 16:32:42 -0700
Subject: Fix bug #78069 - Out-of-bounds read in
 iconv.c:_php_iconv_mime_decode() due to integer overflow

(cherry picked from commit 7cf7148a8f8f4f55fb04de2a517d740bb6253eac)
---
 ext/iconv/iconv.c             |   4 +++-
 ext/iconv/tests/bug78069.data | Bin 0 -> 107 bytes
 ext/iconv/iconv.c             |    4 +++-
 ext/iconv/tests/bug78069.data |binary
 ext/iconv/iconv.c             |    4 +++-
 ext/iconv/tests/bug78069.phpt |   15 +++++++++++++++
 2 files changed, 18 insertions(+), 1 deletion(-)
 create mode 100644 ext/iconv/tests/bug78069.data
 create mode 100644 ext/iconv/tests/bug78069.phpt

--- a/ext/iconv/iconv.c
+++ b/ext/iconv/iconv.c
@@ -1645,7 +1645,9 @@ static php_iconv_err_t _php_iconv_mime_d
 							 * we can do at this point. */
 							if (*(p1 + 1) == '=') {
 								++p1;
-								--str_left;
+								if (str_left > 1) {
+									--str_left;
+								}
 							}
 
 							err = _php_iconv_appendl(pretval, encoded_word, (size_t)((p1 + 1) - encoded_word), cd_pl);
--- /dev/null
+++ b/ext/iconv/tests/bug78069.phpt
@@ -0,0 +1,15 @@
+--TEST--
+Bug #78069 (Out-of-bounds read in iconv.c:_php_iconv_mime_decode() due to integer overflow)
+--SKIPIF--
+<?php
+if (!extension_loaded('iconv')) die('skip ext/iconv required');
+?>
+--FILE--
+<?php
+$hdr = iconv_mime_decode_headers(file_get_contents(__DIR__ . "/bug78069.data"),2);
+var_dump(count($hdr));
+?>
+DONE
+--EXPECT--
+int(1)
+DONE
\ No newline at end of file
